Source/AliveLibAE/Sound/Midi.cpp

```diff
@@ -867,3 +867,220 @@ EXPORT void CC SND_Restart_4CB0E0()
+        if (!ppSeq)
+        {
+            return resource;
+        }
+
+        u32 size = ResourceManager::Get_Header_49C410(ppSeq)->field_0_size;
+        resource->data = *ppSeq;
+        resource->size = size;
+        return resource;
+    }
+
+    s32 sequenceCount()
+    {
+        return 144;
+    }
+
+private:
+    const char_type* mLoadedFile = NULL;
+};
+
+/*
+    AE is different than AO. AE reads sample data from a separate
+    sounds.dat file whereas AO reads it inline of the vab body.
+*/
+class AESoundSampleParser : public psx::SoundSampleParser
+{
+public:
+    std::vector<psx::Sample*> parseSamples(psx::VabHeader* vabHeader, u8* ppVabBody)
+    {
+        std::vector<psx::Sample*> samples;
+        
+
+        std::ifstream soundDatFile;
+        soundDatFile.open("sounds.dat", std::ios::binary);
+        if (!soundDatFile.is_open())
+        {
+            throw std::invalid_argument("Could not find sounds.dat");
+        }
+        soundDatFile.seekg(0, std::ios::end);
+        std::streamsize fileSize = soundDatFile.tellg();
+        soundDatFile.seekg(0, std::ios::beg);
+        char* soundData = new char[(s32) fileSize + 1]; // Plus one, just in case interpolating tries to go that one byte further!
+
+        if (soundDatFile.read(soundData, fileSize))
+        {
+            // Tada!
+        }
+
+        int pos = 0;
+        for (s32 i = 0; i < vabHeader->field_16_num_vags; i++)
+        {
+            u32 size = *reinterpret_cast<u32*>(&ppVabBody[pos]);
+            pos += sizeof(u32);
+
+            u32 sampleRate = *reinterpret_cast<u32*>(&ppVabBody[pos]);
+            pos += sizeof(u32);
+
+            // In AO the body is at this part.
+            // IN AE instead it's an offset in sounds.dat
+            u32 offset = *reinterpret_cast<u32*>(&ppVabBody[pos]);
+            pos += sizeof(u32);
+
+            // Read data from sounds.dat
+            u8* data = new u8[size];
+            memcpy(data, &soundData[offset], size);
+
+            psx::Sample* sample = new psx::Sample();
+            sample->m_SampleBuffer = reinterpret_cast<s16*>(data);
+            sample->i_SampleSize = size / 2;
+            sample->sampleRate = 8000;        // non standard? Doesn't use sampleRate field?
+            sample->loop = sampleRate > 44100; // non-standard?
+            samples.push_back(sample);
+        }
+
+        soundDatFile.close();
+
+        return samples;
+    }
+};
+
+psx::MidiPlayer* player = new psx::MidiPlayer(new AEResourceProvider(), new AESoundSampleParser());
+
+EXPORT void CC SND_StopAll_4CB060()
+{
+    MusicController::EnableMusic_47FE10(FALSE);
+    BackgroundMusic::Stop_4CB000();
+    player->SND_StopAll();
+}
+
+EXPORT void CC SND_Init_4CA1F0()
+{
+    player->SND_Init();
+}
+
+EXPORT void CC SND_Shutdown_4CA280()
+{
+    player->SND_Shutdown();
+}
+
+EXPORT void CC SND_Stop_Channels_Mask_4CA810(u32 bitMask)
+{
+    bitMask;
```

Comment:
This line, and the ones in SND_Load_VABS_4CA350 don't seem to do anything, is this needed?